File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.py
def solution(plans):
    answer = []
    for i in range(len(plans)):
        h, m = map(int, plans[i][1].split(':'))
        st = h*60+m
        plans[i][1] = st
        plans[i][2] = int(plans[i][2])
        
    plans.sort(key=lambda x:x[1])
    
    stack = []
    
    def do_Remain(remain_time): # 다음 숙제까지 남은 시간
        while stack: 
            index, remain_t = stack.pop()
            #print(plans[index], remain_t)
            remain_time -= remain_t
            if remain_time < 0: # 남은거 하다 다음 숙제 해야함
                stack.append((index, remain_time*(-1)))
                break
            else: # 끝냄
                answer.append(plans[index][0])
                #print(plans[index][0])
    
    for i in range(len(plans)):
        if i == len(plans)-1: # 마지막
            answer.append(plans[i][0])
            do_Remain(100000)
            break
        
        # 양수 : 현재 숙제 끝내기 위해 남은 시간 / 음수 : 다음 숙제까지 남은 시간
        remain_time = (plans[i][1] + plans[i][2]) - plans[i+1][1] 
        
        if remain_time > 0: # 시간이 지남
            stack.append((i, remain_time))
        elif remain_time < 0: # 숙제 다하고 남은거 함
            answer.append(plans[i][0])
            do_Remain(remain_time*(-1))
        else: # 숙제끝내고 바로 다음 숙제
            answer.append(plans[i][0])
                   
                
    return answer

File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.java
import java.util.*;

class Solution {
    public String[] solution(String[][] plans) {
        int n = plans.length;
        String[] answer = new String[n];
        
        // PriorityQueue에 o[1] 기준으로 오름차순 
        PriorityQueue<String[]> pq = new PriorityQueue<>((o1, o2) -> {
            String[] time = o1[1].split(":");
            String[] time2 = o2[1].split(":");
            
            int t1 = Integer.parseInt(time[0]) * 60 + Integer.parseInt(time[1]);
            int t2 = Integer.parseInt(time2[0]) * 60 + Integer.parseInt(time2[1]);
            return t1 - t2;
        });
        
        for(String[] plan : plans) {
            pq.offer(plan);
        }
        
        int index = 0;
        Stack<String[]> stack = new Stack<>(); // 중단된 과제 넣는 곳
        String[] now = pq.poll();
        
        while(!pq.isEmpty()) {
            String[] next = pq.poll();
            // System.out.println("now: " + Arrays.toString(now) + " next:" + Arrays.toString(next));
            
            int now_time = calcTime(now[1]);
            int next_time = calcTime(next[1]);
            int diff = next_time - now_time;
            
            int cost = Integer.parseInt(now[2]);
            if(cost > diff) { // 중단
                String remain = Integer.toString(cost - diff);
                stack.push(new String[]{now[0], now[1], remain}); // 남은 시간 push
            } else if (cost == diff) { // 딱 맞게 끝남
                answer[index++] = now[0];
            } else { // cost < diff now 끝나고 next하기까지 시간이 남음
                answer[index++] = now[0];
                
                int remain = diff -cost; // 다음까지 시간
                while(!stack.isEmpty()) {
                    if(remain <= 0) break;
                    String[] s = stack.pop(); // 12:30 90
                    int c = Integer.parseInt(s[2]);
                    if(remain >= c) { // 다음까지 시간 > 남은시간 = answer
                        answer[index++] = s[0];
                    } else { // 다시 stack에 
                        stack.push(new String[]{s[0], s[1], Integer.toString(c-remain)});
                    }
                    remain -= c;
                }
            
            }
            now = next;
            
        }
        // 마지막꺼 넣기
        answer[index++] = now[0];
        
        // stack에 중단된 것들 있으면 넣기
        while(!stack.isEmpty()) {
            String[] s = stack.pop();
            answer[index++] = s[0]; 
        }
        
        
        return answer;
    }
    
    public int calcTime(String time) {
        String[] t = time.split(":");
        
        return Integer.parseInt(t[0]) * 60 + Integer.parseInt(t[1]);
    }
}